### PR TITLE
feat: run tests in 20 groups

### PR DESCRIPTION
**Pull request by faddat**
Running tests in 20 separate groups dramatically accelerates running comet's tests, enhancing the ability to use CI as a single source of truth for tests. 

An additional benefit of this approach is that tests are more isolated from one another.

----

contains changes from #2005 

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments
- [ ] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -31,5 +31,5 @@ jobs:
             Makefile
       - name: Run Go Tests
         run: |
-          make test-group-${{ matrix.part }} NUM_SPLIT=6
+          make test-group-${{ matrix.part }} NUM_SPLIT=20
```
Comment (melekes) [inline .github/workflows/tests.yml:34]: what is the "ideal" split size?

File: Makefile [hunk 2]
```diff
@@ -355,7 +355,7 @@ $(BUILDDIR)/packages.txt:$(GO_TEST_FILES) $(BUILDDIR)
 split-test-packages:$(BUILDDIR)/packages.txt
 	split -d -n l/$(NUM_SPLIT) $< $<.
 test-group-%:split-test-packages
-	cat $(BUILDDIR)/packages.txt.$* | xargs go test -mod=readonly -timeout=15m -race -coverprofile=$(BUILDDIR)/$*.profile.out
+	cat $(BUILDDIR)/packages.txt.$* | xargs go test -mod=readonly -timeout=210s -race -coverprofile=$(BUILDDIR)/$*.profile.out
```
Comment (melekes) [inline Makefile:358]: why 210s?

Comment (faddat) [reply Makefile:358]: Nothing takes longer than that, and so by setting it to 210 seconds, we still allow things to pass while catching race conditions fast

File: .github/workflows/tests.yml (hunk 1 above)
Comment (faddat) [reply .github/workflows/tests.yml:34]: Seemingly 20.  We could go higher but I don't think there's really any need, once we go up to 20 time used is about half tests and half setup.

**Review by melekes — APPROVED** [review]
Thanks @faddat ❤️

**Review by sergio-mena — APPROVED** [review]
🙏 for this. Let's try it out for a few days and see how it works. Let's not backport it for the moment

**Comment by melekes** [comment]
```
panic: test timed out after 3m30s
running tests:
	TestGCRandom (1m18s)

goroutine 9 [running]:
testing.(*M).startAlarm.func1()
	/opt/hostedtoolcache/go/1.21.8/x64/src/testing/testing.go:2259 +0x259
created by time.goFunc
	/opt/hostedtoolcache/go/1.21.8/x64/src/time/sleep.go:176 +0x45

goroutine 1 [chan receive, 1 minutes]:
testing.(*T).Run(0xc000082ea0, {0x71e1ee, 0xc}, 0x7470e0)
	/opt/hostedtoolcache/go/1.21.8/x64/src/testing/testing.go:1649 +0x871
testing.runTests.func1(0x0?)
	/opt/hostedtoolcache/go/1.21.8/x64/src/testing/testing.go:2054 +0x85
testing.tRunner(0xc000082ea0, 0xc000187ae8)
	/opt/hostedtoolcache/go/1.21.8/x64/src/testing/testing.go:1595 +0x262
testing.runTests(0xc0000a2be0?, {0x90ad60, 0x6, 0x6}, {0x1c?, 0x1e?, 0x910e20?})
	/opt/hostedtoolcache/go/1.21.8/x64/src/testing/testing.go:2052 +0x8ae
testing.(*M).Run(0xc0000a2be0)
	/opt/hostedtoolcache/go/1.21.8/x64/src/testing/testing.go:1925 +0xcd8
main.main()
	_testmain.go:97 +0x2e5

goroutine 8 [semacquire]:
sync.runtime_Semacquire(0xc00542c058?)
	/opt/hostedtoolcache/go/1.21.8/x64/src/runtime/sema.go:62 +0x25
sync.(*WaitGroup).Wait(0xc00542c050)
	/opt/hostedtoolcache/go/1.21.8/x64/src/sync/waitgroup.go:[116](https://github.com/cometbft/cometbft/actions/runs/8328441015/job/22788386523?pr=2387#step:5:117) +0xa5
github.com/cometbft/cometbft/internal/clist.TestGCRandom(0xc0000829c0)
	/home/runner/work/cometbft/cometbft/internal/clist/clist_test.go:150 +0x40b
testing.tRunner(0xc0000829c0, 0x7470e0)
	/opt/hostedtoolcache/go/1.21.8/x64/src/testing/testing.go:1595 +0x262
created by testing.(*T).Run in goroutine 1
	/opt/hostedtoolcache/go/1.21.8/x64/src/testing/testing.go:1648 +0x846
FAIL	github.com/cometbft/cometbft/internal/clist	210.176s
ok  	github.com/cometbft/cometbft/internal/cmap	1.014s	coverage: 91.2% of statements
FAIL
make: *** [Makefile:358: test-group-04] Error [123](https://github.com/cometbft/cometbft/actions/runs/8328441015/job/22788386523?pr=2387#step:5:124)
```

keeps failing... I don't know why.

**Comment by faddat** [comment]
is it a new test?

**Comment by melekes** [comment]
I think it was commented out before. I'm not sure when/why it was uncommented.

**Comment by faddat** [comment]
this is odd, let me try giving it a teeny bit more time.

**Comment by faddat** [comment]
Ok, so this is still much faster, but we should probably look into making TestGCRandom run much more quickly.

**Review by melekes — APPROVED** [review]

**Pull request merged**